src/main/kotlin/com/autonomousapps/internal/parse/KotlinBuildScriptDependenciesRewriter.kt

```diff
@@ -0,0 +1,176 @@
+  private val rewriter = Rewriter(tokens)
+  private val indent = Whitespace.computeIndent(tokens, input)
+
+  private val dependencyExtractor = DependencyExtractor(
+    input = input,
+    tokens = tokens,
+    indent = indent,
+  )
+
+  private var hasDependenciesBlock = false
+  private var inBuildscriptBlock = false
+
+  /**
+   * Returns build script with advice applied.
+   *
+   * Throws [BuildScriptParseException] if the script has some idiosyncrasy that impairs parsing.
+   *
+   */
+  @Throws(BuildScriptParseException::class)
+  override fun rewritten(): String {
+    errorListener.getErrorMessages().ifNotEmpty {
+      throw BuildScriptParseException.withErrors(it)
+    }
+
+    return rewriter.text
+  }
+
+  override fun enterNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onEnterBlock()
+
+    if (ctx.isBuildscript) {
+      inBuildscriptBlock = true
+    }
+
+    if (ctx.isDependencies) {
+      handleDependencies(ctx)
+    }
+  }
+
+  override fun exitNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onExitBlock()
+
+    if (ctx.isDependencies && !inBuildscriptBlock) {
+      hasDependenciesBlock = true
+      advice.filterToSet { it.isAnyAdd() }.ifNotEmpty { addAdvice ->
+        val closeBrace = ctx.stop
+        rewriter.insertBefore(closeBrace, addAdvice.joinToString(separator = "\n", postfix = "\n") { a ->
+          printer.toDeclaration(a)
+        })
+      }
+    }
+
+    // Must be last
+    if (ctx.isBuildscript) {
+      inBuildscriptBlock = false
+    }
+  }
+
+  override fun exitScript(ctx: ScriptContext) {
+    // Exit early if this build script has a dependencies block. If it doesn't, we may need to add missing dependencies.
+    if (hasDependenciesBlock) return
+
+    advice.filterToOrderedSet { it.isAnyAdd() }.ifNotEmpty { addAdvice ->
+      rewriter.insertBefore(
+        ctx.stop,
+        addAdvice.joinToString(prefix = "\ndependencies {\n", postfix = "\n}\n", separator = "\n") { a ->
+          printer.toDeclaration(a)
+        }
+      )
+    }
+  }
+
+  private fun handleDependencies(ctx: NamedBlockContext) {
+    if (inBuildscriptBlock) return
+
+    val dependencyContainer = dependencyExtractor.collectDependencies(ctx)
+    dependencyContainer.getDependencyDeclarationsWithContext().forEach {
+      if (it.statement.leafRule() !is PostfixUnaryExpressionContext) return@forEach
+
+      val context = it.statement.leafRule() as PostfixUnaryExpressionContext
+      val declaration = it.declaration
+
+      findAdvice(declaration)?.let { a ->
+        if (a.isAnyRemove()) {
+          rewriter.delete(context.start, context.stop)
+          rewriter.deleteWhitespaceToLeft(context.start)
+          rewriter.deleteNewlineToRight(context.stop)
+        } else if (a.isAnyChange()) {
+          val configuration = context.primaryExpression()
+          rewriter.replace(configuration.start, configuration.stop, a.toConfiguration)
+        }
```

Comment:
I think this is being too clever. The `DependencyDeclaration` data class has all the information we need, and we shouldn't try to re-parse the context object to get the configuration context out of it. We should just rewrite the whole line using `printer.toDeclaration(a)`. That ensures there's just one method responsible for printing declarations.

I think the above ☝️ is true, but I will note one caveat that came to mind as I wrote it. There are cases where there are really complex dependency declarations, and what I said assumes that DAGP is correctly modeling a declaration in its totality. Your approach is more conservative by _just_ rewriting the configuration name. 

All that said, I prefer keeping the code simple and using `printer.toDeclaration(a)` to rewrite the whole line. `fixDependencies` is an optional task and is allowed to make assumptions that `buildHealth` can't.

Reply:
> . We should just rewrite the whole line using printer.toDeclaration(a). That ensures there's just one method responsible for printing declarations.

Good point. I will update this

Reply:
👍 I'm happy to update this to `printer.toDeclaration(a)` for simplicity. I was just following the existing pattern e.g. https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/d36ba8a9cf5cf6d03ba51863f32512ff38002ced/src/main/kotlin/com/autonomousapps/internal/parse/GroovyBuildScriptDependenciesRewriter.kt#L149-L150

### Note on the test set up
After making the [change](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/7c7c9154d8885437092bd881e34637860ae1fc87/src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt), the test "update dependencies with dependencyMap" test failed e.g. `compileOnly(project(:marvin))` (without `""`) instead of `compileOnly(project(":marvin"))`

It's because the `AdvicePrinter` doesn't print the quote if `dependencyMap` is presented for the specific advice.
https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/7c7c9154d8885437092bd881e34637860ae1fc87/src/main/kotlin/com/autonomousapps/internal/advice/AdvicePrinter.kt#L58-L66

To get the test passing, I need to avoid providing `dependencyMap` for the dependency we want to remain the same. See [change](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1299/commits/c358cd6521cc6afbfa677514d6b275be42d16e10#diff-85ff32ed71048450273d2fd46bcc842a7ef269ee87773608f362a8a55cc1bb96L152)  

### Question
Can we assume that the given `dependencyMap` from users should always include quotes?

Reply:
> Can we assume that the given dependencyMap from users should always include quotes?

Yes. Good catch and fix!